docs/tutorial/t-spark-monitoring.md

```diff
@@ -0,0 +1,350 @@
+alertmanager:grafana-dashboard      grafana:grafana-dashboard    grafana_dashboard      regular  
+alertmanager:grafana-source         grafana:grafana-source       grafana_datasource     regular  
+alertmanager:replicas               alertmanager:replicas        alertmanager_replica   peer     
+alertmanager:self-metrics-endpoint  prometheus:metrics-endpoint  prometheus_scrape      regular  
+catalogue:catalogue                 alertmanager:catalogue       catalogue              regular  
+catalogue:catalogue                 grafana:catalogue            catalogue              regular  
+catalogue:catalogue                 prometheus:catalogue         catalogue              regular  
+catalogue:replicas                  catalogue:replicas           catalogue_replica      peer     
+grafana:grafana                     grafana:grafana              grafana_peers          peer     
+grafana:metrics-endpoint            prometheus:metrics-endpoint  prometheus_scrape      regular  
+grafana:replicas                    grafana:replicas             grafana_replicas       peer     
+loki:grafana-dashboard              grafana:grafana-dashboard    grafana_dashboard      regular  
+loki:grafana-source                 grafana:grafana-source       grafana_datasource     regular  
+loki:metrics-endpoint               prometheus:metrics-endpoint  prometheus_scrape      regular  
+loki:replicas                       loki:replicas                loki_replica           peer     
+prometheus:grafana-dashboard        grafana:grafana-dashboard    grafana_dashboard      regular  
+prometheus:grafana-source           grafana:grafana-source       grafana_datasource     regular  
+prometheus:prometheus-peers         prometheus:prometheus-peers  prometheus_peers       peer     
+traefik:ingress                     alertmanager:ingress         ingress                regular  
+traefik:ingress                     catalogue:ingress            ingress                regular  
+traefik:ingress-per-unit            loki:ingress                 ingress_per_unit       regular  
+traefik:ingress-per-unit            prometheus:ingress           ingress_per_unit       regular  
+traefik:metrics-endpoint            prometheus:metrics-endpoint  prometheus_scrape      regular  
+traefik:peers                       traefik:peers                traefik_peers          peer     
+traefik:traefik-route               grafana:ingress              traefik_route          regular  
+```
+
+At this point, the observability stack has been deployed but Charmed Spark knows nothing about it. Since Charmed Spark comes with a built in spark metrics exporter that exports metric logs to a Prometheus gateway, we need to deploy a Prometheus gateway and then add Spark options to connect to it. The Prometheus gateway in turn will then be integrated with Prometheus. Let's deploy `prometheus-pushgateway-k8s` charm and integrate it with `prometheus` charm.
```

Comment:
```suggestion
At this point, the observability stack has been deployed but Charmed Spark is not wired to it yet. Generally, Prometheus collects metrics of services by regularly scraping dedicated endpoints. However, Spark jobs can be ephemeral processes that may last not so long to be scraped. Moreover, the IPs/hostnames of pods are likely to change among runs, therefore requiring a self-discoverable automation. This is the reason why we decided to provide support for jobs run with Charmed Spark to push metrics (rather than having Prometheus pulling them) to a caching service, called Prometheus pushgateway, that stores the metrics and exposed them to Prometheus for regular scraping, even when the Spark job has finished. Therefore, to wire Spark jobs with the observability stack, we need to deploy a Prometheus pushgateway and then add Spark options to connect to it. The Prometheus pushgateway in turn will then be integrated with Prometheus. Let's deploy `prometheus-pushgateway-k8s` charm and integrate it with `prometheus` charm.
```